dreem/models/transformer.py

```diff
@@ -505,31 +511,40 @@ def forward(
         self,
         decoder_queries: torch.Tensor,
         encoder_features: torch.Tensor,
-        ref_pos_emb: torch.Tensor | None = None,
-        query_pos_emb: torch.Tensor | None = None,
+        embedding_map: Dict[str, Embedding],
+        enc_boxes: torch.Tensor, enc_times: torch.Tensor,
+        boxes: torch.Tensor, times: torch.Tensor,
+        embedding_agg_method: str = None
     ) -> torch.Tensor:
         """Execute a forward pass of the decoder block.
 
         Args:
             decoder_queries: Query sequence for decoder to generate (n_query, batch_size, embed_dim).
             encoder_features: Output from encoder, that decoder uses to attend to relevant
                 parts of input sequence (total_instances, batch_size, embed_dim)
-            ref_pos_emb: The input positional embedding tensor of shape (total_instances, batch_size, embed_dim).
-            query_pos_emb: The query positional embedding of shape (n_query, batch_size, embed_dim)
+            
 
         Returns:
             The output tensor of shape (L, n_query, batch_size, embed_dim).
         """
         decoder_features = decoder_queries
-
         intermediate = []
 
+        # since the encoder output doesn't change for any number of decoder layer inputs, 
+        # we can process its embedding outside the loop
+        if embedding_agg_method == "average":
+            encoder_features, *_ = apply_embeddings(encoder_features, embedding_map,
+                                                enc_boxes, enc_times, embedding_agg_method)
+            # TODO: ^ should embeddings really be applied to encoder output again before cross attention?
+            #   switched off for stack and concatenate methods as those further split the tokens. Kept for "average"
+            #   for backward compatibility
```

Comment:
My understanding is that repeated embedding injection is helpful to maintain the desired positional inductive bias, but probably empirical question. Thanks for maintaining backwards compat though!